proposals/299-ip-failure-count.txt

```diff
@@ -156,20 +156,24 @@ Ticket: https://trac.torproject.org/projects/tor/ticket/27491
    One way we could forget old sessions is by halving all the failure point
    (FP) values before adding when:
 
-      * One or more failure point values are a multiple of a random number
-        between 1 and 5
+      * The failure point value being added to is a multiple of a random
+        number between 1 and 5
```

Comment:
I don't understand why the test is "a multiple of a random number between 1 and 5". And I'm not sure I know how to implement that test in C.

It's also not clear to me how often the test is true, so it is hard to work out if we are reaching our goal of keeping around 30 seconds of history.

Can you please explain why you chose this test?
Can you explain how often you expect it to be true?

Reply:
>Can you please explain why you chose this test?

I initially chose random numbers because we can sometimes halve. But you made me realize that it may not be that often after all.

>Can you explain how often you expect it to be true?

Probably 1/5 of the time. But looking at your other comments, it may not be enough for mobile clients.

Reply:
I am proposing that instead of halving at a multiple of a random number, we halve if the affinity point value is a multiple of 2. This happens if the affinity point values have already been seeded (even if they have already been halved in the past). This helps accommodate mobile clients which could change networks.

Reply:
Or maybe we should halve when we add.

Reply:
Ok, let's think about each of the possible designs where we halve when adding, starting with an affinity point value of zero:

If we always halve before adding:
0/2 + 1 = 1
0/2 + 2 = 2
1/2 + 1 = 1 
1/2 + 2 = 2
The value one is meaningless, because the result of any addition is the same as adding to zero.

If we always halve after adding:
(0 + 1)/2 = 0
Adding one to zero meaningless - the result is zero.
(0 + 2)/2 = 1
(1 + 1)/ 2 = 1
(1 + 2)/2 = 1
Adding two to one is meaningless - the result is the same as adding one to one.

If we halve before adding, but only when the old value is divisible by two:
0/2 + 1 = 1
0/2 + 2 = 2
1 + 1 = 2
1 + 2 = 3
3 + 2 = 5
If you start with an odd number, and only get unreachable, then it never halves.

If we halve after adding, but only when the new value is divisible by two:
(0 + 1) = 1
(0 + 2)/2 = 1
Adding one or two produces the same result. That's probably not what we want.
(1 + 1)/ 2 = 1
(1 + 2) = 3

I don't think halving based on the values is a good idea. We will end up with weird number patterns. Instead, let's choose a random number of connections between 0 and N, and halve after we've seen that many connections. We can use our fast PRNG, so the random number generation will be fast.

How often do we want to halve the values?

We want to keep history from the last few connections.

The summarised affinity point values use a multiple of 4, so we want to have affinity values that are at least 4. If we add at least one point to one counter every connection, and halve both counters every 16 connections, then the values will converge on 4. (1*16/2/2=4) So 16 is the minimum halving interval.

So let's choose a random number between 8 and 24 (8+24/2=16), and halve both values after we've seen that many connections. And then let's choose another random number.

Reply:
Your design is great. I will go with this.